moveit_core/collision_detection_bullet/src/collision_env_bullet.cpp

```diff
@@ -208,27 +209,65 @@ void CollisionEnvBullet::checkRobotCollisionHelperCCD(const CollisionRequest& re
 
+  // Create collision objects for all attached objects
   std::vector<collision_detection_bullet::CollisionObjectWrapperPtr> attached_cows;
-  addAttachedOjects(state1, attached_cows);
+  createAttachedCollisionObjects(state1, attached_cows);
 
+  // Add those collision objects to the CCD collision manager.
   for (const collision_detection_bullet::CollisionObjectWrapperPtr& cow : attached_cows)
   {
     manager_CCD_->addCollisionObject(cow);
-    manager_CCD_->setCastCollisionObjectsTransform(
-        cow->getName(), state1.getAttachedBody(cow->getName())->getGlobalCollisionBodyTransforms()[0],
-        state2.getAttachedBody(cow->getName())->getGlobalCollisionBodyTransforms()[0]);
   }
 
-  for (const std::string& link : active_)
+  // Compute the number of sub-steps (segments in the motion).
+  size_t num_segments = std::max((size_t)std::ceil(state1.maximumRotation(state2) / max_rotation_per_step_), (size_t)1);
+
+  // Compute the intermediate states through interpolation (number of segments - 1)
+  std::vector<moveit::core::RobotState> intermediate_states;
+  intermediate_states.reserve(num_segments - 1);
+  for (size_t segment_idx = 1; segment_idx < num_segments; ++segment_idx)
   {
-    manager_CCD_->setCastCollisionObjectsTransform(link, state1.getCollisionBodyTransform(link, 0),
-                                                   state2.getCollisionBodyTransform(link, 0));
+    double t = (double)segment_idx / (double)num_segments;
+    moveit::core::RobotState st(state1);
+    state1.interpolate(state2, t, st);
+    st.update(true);
+    intermediate_states.push_back(st);
   }
 
-  manager_CCD_->contactTest(res, req, acm, false);
+  // Go through the segments one-by-one.
+  for (size_t segment_idx = 0; segment_idx < num_segments; ++segment_idx)
+  {
+    // Look up the start-and-end states of that segment.
+    const moveit::core::RobotState& st1 = segment_idx == 0 ? state1 : intermediate_states[segment_idx - 1];
+    const moveit::core::RobotState& st2 = segment_idx == (num_segments - 1) ? state2 : intermediate_states[segment_idx];
+
+    // Update the transforms, first for the attached objects...
+    for (const collision_detection_bullet::CollisionObjectWrapperPtr& cow : attached_cows)
+    {
+      manager_CCD_->setCastCollisionObjectsTransform(
+          cow->getName(), st1.getAttachedBody(cow->getName())->getGlobalCollisionBodyTransforms()[0],
+          st2.getAttachedBody(cow->getName())->getGlobalCollisionBodyTransforms()[0]);
+    }
+
+    // ...and then the links themselves.
+    for (const std::string& link : active_)
+    {
+      manager_CCD_->setCastCollisionObjectsTransform(link, st1.getCollisionBodyTransform(link, 0),
+                                                     st2.getCollisionBodyTransform(link, 0));
+    }
+
+    // Perform the contact test
+    manager_CCD_->contactTest(res, req, acm, false);
+
+    // Stop the operation after the first collision.
+    if (res.collision)
```

Comment:
you should actually respect `req.max_contacts` here and keep going until you found enough contacts.
That will probably also fix your invalidated CI test because you miss collisions that only happen further into the trajectory.

To abort checking on collision, users can do checks for just one contact.